internal/controller/apiextensions/composite/composed.go

```diff
@@ -54,6 +60,224 @@ const (
 	errFmtConnDetailPath = "connection detail of type %q fromFieldPath is not set"
 )
 
+// Annotation keys.
+const (
+	AnnotationKeyCompositionTemplateName = "crossplane.io/composition-template-name"
+)
+
+// SetCompositionTemplateName sets the name of the composition template used to
+// reconcile a composed resource as an annotation.
+func SetCompositionTemplateName(o metav1.Object, name string) {
+	meta.AddAnnotations(o, map[string]string{AnnotationKeyCompositionTemplateName: name})
+}
+
+// GetCompositionTemplateName gets the name of the composition template used to
+// reconcile a composed resource from its annotations.
+func GetCompositionTemplateName(o metav1.Object) string {
+	return o.GetAnnotations()[AnnotationKeyCompositionTemplateName]
+}
+
+// A CompositionValidator validates the supplied Composition.
+type CompositionValidator interface {
+	Validate(comp *v1.Composition) error
+}
+
+// A CompositionValidatorFn validates the supplied Composition.
+type CompositionValidatorFn func(comp *v1.Composition) error
+
+// Validate the supplied Composition.
+func (fn CompositionValidatorFn) Validate(comp *v1.Composition) error {
+	return fn(comp)
+}
+
+// A ValidationChain runs multiple validations.
+type ValidationChain []CompositionValidator
+
+// Validate the supplied Composition.
+func (vs ValidationChain) Validate(comp *v1.Composition) error {
+	for _, v := range vs {
+		if err := v.Validate(comp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RejectMixedTemplates validates that the supplied Composition does not attempt
+// to mix named and anonymous templates. If some but not all templates are named
+// it's safest to refuse to operate. We don't have enough information to use the
+// named composer, but using the anonymous composer may be surprising. There's a
+// risk that someone added a new anonymous template to a Composition that
+// otherwise uses named templates. If they added the new template to the
+// beginning or middle of the resources array using the anonymous composer would
+// be destructive, because it assumes template N always corresponds to existing
+// template N.
+func RejectMixedTemplates(comp *v1.Composition) error {
+	named := 0
+	for _, tmpl := range comp.Spec.Resources {
+		if tmpl.TemplateName != nil {
+			named++
+		}
+	}
+
+	// We're using only anonymous templates.
+	if named == 0 {
+		return nil
+	}
+
+	// We're using only named templates.
+	if named == len(comp.Spec.Resources) {
+		return nil
+	}
+
+	return errors.New(errMixed)
+}
+
+// RejectDuplicateNames validates that all template names are unique within the
+// supplied Composition.
+func RejectDuplicateNames(comp *v1.Composition) error {
+	seen := map[string]bool{}
+	for _, tmpl := range comp.Spec.Resources {
+		if tmpl.TemplateName == nil {
+			continue
```

Comment:
Yet another option could be to have a wrapping `CompositionValidatorFn` called `TemplateValidator(TemplateValidatorFn...)` that these functions are passed into and it just calls them in the `comp.Spec.Resources` loop. I don't have a great feel on the potential for this package to expand to more validation, so this might be way over-optimizing at this point. This pattern is similar to the package linter, which I went ahead and made this optimization in because I estimated that a potentially very large number of linters could be passed, so I wanted something more granular than just a package linter func (hence the object and meta linters) https://github.com/crossplane/crossplane-runtime/blob/45bc6d50351e8e12b1da7ca21281d754eb76fb09/pkg/parser/linter.go#L72. This package is also internal so it won't be consumed by others downstream anywhere so less risk in keeping it simple for now.

Reply:
I'd prefer not to break or return here. I like that the functionality this validator declares is "self contained" - i.e. it will reject a template with duplicate template names regardless of whether or not all the templates are named. We can come back and address whether these validation functions could be improved when we migrate validation to a webhook.

Reply:
That’s fair. I think my primary concern was not returning an error when determining uniqueness because of a mix of named and anonymous, but this function is not actually really saying that, it is only saying that the names do not conflict, which is still true in this case 👍